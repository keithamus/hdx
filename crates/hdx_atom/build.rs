use std::{collections::HashSet, env, path::Path};

use glob::glob;
use grep_matcher::{Captures, Matcher};
use grep_regex::RegexMatcher;
use grep_searcher::{sinks::UTF8, Searcher};

pub fn kebab(str: String) -> String {
	let mut kebab = String::new();
	for (i, ch) in str.char_indices() {
		if i > 0 && ch.is_uppercase() {
			kebab.push('-');
		}
		kebab.push(ch.to_ascii_lowercase());
	}
	kebab
}

fn main() {
	println!("cargo::rerun-if-changed=build.rs");

	let matcher = RegexMatcher::new_line_matcher(r#"(atom!\("|#\[value\(")([^"\)]+)(?: "\)\])?"#).unwrap();
	let mut matches = HashSet::new();
	matches.insert("%".to_owned());
	for entry in glob("../**/*.rs").unwrap() {
		// for entry in glob("../**/values/ui/mod.rs").unwrap() {
		let str = &entry.as_ref().unwrap().display();
		println!("cargo::rerun-if-changed={}", str);
		let mut searcher = Searcher::new();
		searcher
			.search_path(
				&matcher,
				entry.unwrap(),
				UTF8(|_lnum, line| {
					let mut captures = matcher.new_captures()?;
					matcher.captures_iter(line.as_bytes(), &mut captures, |captures| -> bool {
						let start = &line[captures.get(1).unwrap()];
						let capture = &line[captures.get(2).unwrap()];
						if start == "#[value(\"" {
							let keywords = capture
								.split(" | ")
								.map(|part| {
									part.trim()
										.split(" ")
										.next()
										.unwrap_or("")
										.trim_start_matches('[')
										.trim_end_matches(']')
										.trim_end_matches('?')
										.trim_end_matches('(')
										.trim_start_matches(' ')
										.trim()
								})
								.filter(|part| !(part.is_empty() || part.starts_with('<')))
								.collect::<Vec<&str>>();
							for keyword in keywords {
								if keyword.chars().all(|c| c == '-' || char::is_alphanumeric(c)) {
									matches.insert(keyword.to_owned());
								}
							}
						} else if capture.chars().all(|c| c == '-' || c == '_' || char::is_alphanumeric(c)) {
							matches.insert(capture.to_owned());
						}
						true
					})?;
					Ok(true)
				}),
			)
			.unwrap();
	}

	let matcher = RegexMatcher::new_line_matcher("^pub (?:struct|enum) (\\w*)").unwrap();
	for entry in glob("../hdx_ast/src/css/values/*/mod.rs").unwrap() {
		let str = &entry.as_ref().unwrap().display();
		println!("cargo::rerun-if-changed={}", str);
		let mut searcher = Searcher::new();
		searcher
			.search_path(
				&matcher,
				entry.unwrap(),
				UTF8(|_lnum, line| {
					let mut captures = matcher.new_captures()?;
					matcher.captures_iter(line.as_bytes(), &mut captures, |captures| -> bool {
						let capture = &line[captures.get(1).unwrap()];
						if !capture.is_empty() {
							let mut atom_name = kebab(capture.trim_end_matches("StyleValue").into());
							if atom_name.starts_with("webkit") {
								atom_name = format!("-{}", atom_name);
							}
							matches.insert(atom_name);
						}
						true
					})?;
					Ok(true)
				}),
			)
			.unwrap();
	}

	string_cache_codegen::AtomType::new("Atom", "atom!")
		.atoms(matches)
		.write_to_file(&Path::new(&env::var("OUT_DIR").unwrap()).join("hdx_atom.rs"))
		.unwrap();
}
