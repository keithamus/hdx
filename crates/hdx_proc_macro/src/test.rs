use crate::{def::*, value::generate};
use hdx_atom::{atom, Atom};
use quote::quote;

macro_rules! to_valuedef {
	( $lit:literal ) => {
		::syn::parse2::<StrWrapped<Def>>(::quote::quote!{ $lit }).unwrap().0
	};
	( $($tt:tt)+ ) => {
		::syn::parse2::<Def>(::quote::quote!{ $($tt)+ }).unwrap()
	};
}

macro_rules! to_deriveinput {
	( $($tt:tt)+ ) => {
		::syn::parse2::<::syn::DeriveInput>(::quote::quote!{ $($tt)+ }).unwrap()
	}
}

macro_rules! assert_snapshot {
	( $syntax:ident, $data:ident, $name:literal) => {
		let file = ::syn::parse2::<syn::File>(generate($syntax, $data)).unwrap();
		let pretty = ::prettyplease::unparse(&file);
		::insta::assert_snapshot!($name, pretty)
	};
}

#[test]
fn test_def_builds_type() {
	assert_eq!(to_valuedef!( <integer> ), Def::Type(DefType::Integer(DefRange::None)))
}

#[test]
fn test_def_builds_quoted_type() {
	assert_eq!(
		::syn::parse2::<StrWrapped<Def>>(quote! { "<'some-prop'>" }).unwrap().0,
		Def::Type(DefType::Custom(
			DefIdent(Atom::from("SomePropStyleValue")),
			DefIdent(Atom::from("SomePropStyleValue"))
		))
	)
}

#[test]
fn test_def_builds_type_with_multiplier_oneormore() {
	assert_eq!(
		to_valuedef!( <integer>+ ),
		Def::Multiplier(Box::new(Def::Type(DefType::Integer(DefRange::None))), DefMultiplierStyle::OneOrMore)
	)
}

#[test]
fn def_builds_type_with_checks() {
	assert_eq!(to_valuedef! { <integer [1,3]> }, Def::Type(DefType::Integer(DefRange::Range(1f32..3f32))))
}

#[test]
fn test_def_builds_optional() {
	assert_eq!(to_valuedef!( <integer>? ), Def::Optional(Box::new(Def::Type(DefType::Integer(DefRange::None)))))
}

#[test]
fn test_def_builds_quoted_custom_type_with_count() {
	assert_eq!(
		::syn::parse2::<StrWrapped<Def>>(quote! { "<'animation-delay'>{1,3}" }).unwrap().0,
		Def::Multiplier(
			Box::new(Def::Type(DefType::Custom(
				DefIdent(atom!("AnimationDelayStyleValue")),
				DefIdent(atom!("AnimationDelayStyleValue"))
			))),
			DefMultiplierStyle::Range(DefRange::Range(1.0..3.0))
		)
	)
}

#[test]
fn def_builds_combinator_of_keywords() {
	assert_eq!(
		to_valuedef! { none | auto },
		Def::Combinator(
			vec![Def::Ident(DefIdent(atom!("none"))), Def::Ident(DefIdent(atom!("auto")))],
			DefCombinatorStyle::Alternatives,
		)
	)
}

#[test]
fn def_builds_ordered_combinator_of_keywords() {
	assert_eq!(
		to_valuedef! { none auto },
		Def::Combinator(
			vec![Def::Ident(DefIdent(atom!("none"))), Def::Ident(DefIdent(atom!("auto")))],
			DefCombinatorStyle::Ordered,
		)
	)
}

#[test]
fn test_def_builds_dashed_idents() {
	assert_eq!(
		to_valuedef!( length-percentage preserve-3d  ),
		Def::Combinator(
			vec![Def::Ident(DefIdent(atom!("length-percentage"))), Def::Ident(DefIdent(atom!("preserve-3d")))],
			DefCombinatorStyle::Ordered,
		)
	)
}

#[test]
fn def_builds_group_with_brackets() {
	assert_eq!(
		to_valuedef! { [ block || inline ] | none },
		Def::Combinator(
			vec![
				Def::Group(
					Box::new(Def::Combinator(
						vec![Def::Ident(DefIdent(atom!("block"))), Def::Ident(DefIdent(atom!("inline")))],
						DefCombinatorStyle::Options,
					)),
					DefGroupStyle::None,
				),
				Def::Ident(DefIdent(atom!("none"))),
			],
			DefCombinatorStyle::Alternatives,
		)
	);
}

#[test]
fn def_builds_combinator_with_correct_precedence() {
	assert_eq!(
		to_valuedef! { none | underline || overline },
		Def::Combinator(
			vec![
				Def::Ident(DefIdent(atom!("none"))),
				Def::Combinator(
					vec![Def::Ident(DefIdent(atom!("underline"))), Def::Ident(DefIdent(atom!("overline")))],
					DefCombinatorStyle::Options,
				),
			],
			DefCombinatorStyle::Alternatives,
		)
	);
}

#[test]
fn def_builds_combinator_with_correct_precedence2() {
	assert_eq!(
		to_valuedef! { underline || overline | none },
		Def::Combinator(
			vec![
				Def::Combinator(
					vec![Def::Ident(DefIdent(atom!("underline"))), Def::Ident(DefIdent(atom!("overline")))],
					DefCombinatorStyle::Options,
				),
				Def::Ident(DefIdent(atom!("none"))),
			],
			DefCombinatorStyle::Alternatives,
		)
	);
}

#[test]
fn def_builds_combinator_with_correct_precedence3() {
	assert_eq!(
		to_valuedef! { auto none | underline || overline && block inline },
		Def::Combinator(
			vec![
				Def::Combinator(
					vec![Def::Ident(DefIdent(atom!("auto"))), Def::Ident(DefIdent(atom!("none")))],
					DefCombinatorStyle::Ordered,
				),
				Def::Combinator(
					vec![
						Def::Ident(DefIdent(atom!("underline"))),
						Def::Combinator(
							vec![
								Def::Ident(DefIdent(atom!("overline"))),
								Def::Combinator(
									vec![Def::Ident(DefIdent(atom!("block"))), Def::Ident(DefIdent(atom!("inline")))],
									DefCombinatorStyle::Ordered,
								),
							],
							DefCombinatorStyle::AllMustOccur,
						),
					],
					DefCombinatorStyle::Options,
				),
			],
			DefCombinatorStyle::Alternatives,
		)
	);
}

#[test]
fn def_builds_group_of_types_and_keywords() {
	assert_eq!(
		to_valuedef! { <length [1,]> | auto },
		Def::Combinator(
			vec![Def::Type(DefType::Length(DefRange::RangeFrom(1f32..))), Def::Ident(DefIdent(atom!("auto")))],
			DefCombinatorStyle::Alternatives,
		)
	)
}

#[test]
fn def_builds_multiplier_of_types() {
	assert_eq!(
		to_valuedef! { <length># },
		Def::Multiplier(
			Box::new(Def::Type(DefType::Length(DefRange::None))),
			DefMultiplierStyle::OneOrMoreCommaSeparated(DefRange::None)
		)
	)
}

#[test]
fn def_builds_with_literal_chars() {
	assert_eq!(
		to_valuedef! { <color> / <color> },
		Def::Combinator(
			vec![Def::Type(DefType::Color), Def::Punct('/'), Def::Type(DefType::Color)],
			DefCombinatorStyle::Ordered,
		)
	)
}

#[test]
fn def_builds_multiplier_of_types_with_range() {
	let range = 5f32..12f32;
	assert_eq!(
		to_valuedef! { <length>#{5,12} },
		Def::Multiplier(
			Box::new(Def::Type(DefType::Length(DefRange::None))),
			DefMultiplierStyle::OneOrMoreCommaSeparated(DefRange::Range(range))
		)
	)
}

#[test]
fn def_builds_multiplier_of_type_fixed_range() {
	assert_eq!(
		to_valuedef! { <length>{5} },
		Def::Multiplier(
			Box::new(Def::Type(DefType::Length(DefRange::None))),
			DefMultiplierStyle::Range(DefRange::Fixed(5f32))
		)
	)
}

#[test]
fn def_builds_complex_combination_1() {
	assert_eq!(
		to_valuedef! { [ inset? && <length>{2,4} && <color>? ]# | none },
		Def::Combinator(
			vec![
				Def::Group(
					Box::new(Def::Combinator(
						vec![
							Def::Optional(Box::new(Def::Ident(DefIdent(atom!("inset"))))),
							Def::Multiplier(
								Box::new(Def::Type(DefType::Length(DefRange::None))),
								DefMultiplierStyle::Range(DefRange::Range(2f32..4f32)),
							),
							Def::Optional(Box::new(Def::Type(DefType::Color))),
						],
						DefCombinatorStyle::AllMustOccur,
					)),
					DefGroupStyle::OneOrMore,
				),
				Def::Ident(DefIdent(atom!("none"))),
			],
			DefCombinatorStyle::Alternatives,
		)
	)
}

#[test]
fn value_lone_type() {
	let syntax = to_valuedef! { <integer> };
	let data = to_deriveinput! { struct Foo; };
	assert_snapshot!(syntax, data, "value_lone_type");
}

#[test]
fn value_lone_type_errors_with_lifetime_when_necessary() {
	let syntax = to_valuedef! { <image> }; // <image> needs lifetime
	let data = to_deriveinput! { struct Foo; }; // Foo has no lifetime
	assert_snapshot!(syntax, data, "value_lone_type_errors_with_lifetime_when_necessary");
}

#[test]
fn value_lone_type_with_lifetime_2() {
	let syntax = to_valuedef! { <image> }; // <image> needs lifetime
	let data = to_deriveinput! { struct Foo<'a>; }; // Foo specifies lifetime
	assert_snapshot!(syntax, data, "value_lone_type_with_lifetime");
}

#[test]
fn value_lone_custom_type() {
	let syntax = to_valuedef! { <custom-ident> };
	let data = to_deriveinput! { struct Foo; };
	assert_snapshot!(syntax, data, "value_lone_custom_type");
}

#[test]
fn enum_type_with_lifetime() {
	let syntax = to_valuedef! { <color> | <image-1D> }; // <image-1D> needs lifetime
	let data = to_deriveinput! { enum Foo<'a> {} }; // Foo specifies lifetime
	assert_snapshot!(syntax, data, "enum_type_with_lifetime");
}

#[test]
fn multiple_keywords() {
	let syntax = to_valuedef!("black | white | line-through | pink");
	let data = to_deriveinput! { enum Foo {} };
	assert_snapshot!(syntax, data, "multiple_keywords");
}

#[test]
fn value_group_type_keyword() {
	let syntax = to_valuedef!( <length [1,]> | line-through );
	let data = to_deriveinput! { enum Foo {} };
	assert_snapshot!(syntax, data, "value_group_type_keyword");
}

#[test]
fn value_with_multiplier_range() {
	let syntax = to_valuedef!( <length>{2,4} );
	let data = to_deriveinput! { struct Foo; };
	assert_snapshot!(syntax, data, "value_with_multiplier_range");
}

#[test]
fn keyword_or_type() {
	let syntax = to_valuedef!( none | <custom-ident> );
	let data = to_deriveinput! { enum Foo {} };
	assert_snapshot!(syntax, data, "keyword_or_type");
}

#[test]
fn custom_type_with_checks() {
	let syntax = to_valuedef!(" none | <length-percentage [0,∞]> ");
	let data = to_deriveinput! { enum Foo {} };
	assert_snapshot!(syntax, data, "custom_type_with_checks");
}

#[test]
fn custom_function_type() {
	let syntax = to_valuedef!(" none | <calc-size()> ");
	let data = to_deriveinput! { enum Foo {} };
	assert_snapshot!(syntax, data, "custom_function_type");
}

#[test]
fn custom_function_variant_with_args() {
	let syntax = to_valuedef!(" fit-content | fit-content(<length-percentage [0,∞]>) ");
	let data = to_deriveinput! { enum Foo {} };
	assert_snapshot!(syntax, data, "custom_function_variant_with_args");
}

#[test]
fn custom_function_variant_with_multiplier_args() {
	let syntax = to_valuedef!(" normal | styleset(<feature-value-name>#) ");
	let data = to_deriveinput! { enum Foo<'a> {} };
	assert_snapshot!(syntax, data, "custom_function_variant_with_multiplier_args");
}

#[test]
fn custom_function_all_optionals() {
	let syntax = to_valuedef!(" <'caret-color'> || <'caret-animation'> || <'caret-shape'> ");
	let data = to_deriveinput! { struct Foo; };
	assert_snapshot!(syntax, data, "custom_function_all_optionals");
}

#[test]
fn ordered_custom_function_last_option() {
	let syntax = to_valuedef!(" <'caret-color'> <'caret-animation'>? ");
	let data = to_deriveinput! { struct Foo; };
	assert_snapshot!(syntax, data, "ordered_custom_function_last_option");
}

#[test]
fn struct_with_variable_count_type() {
	let syntax = to_valuedef!(" <animateable-feature># ");
	let data = to_deriveinput! { struct Foo<'a>; };
	assert_snapshot!(syntax, data, "struct_with_variable_count_type");
}

#[test]
fn enum_with_variable_count_type() {
	let syntax = to_valuedef!(" auto | <animateable-feature># ");
	let data = to_deriveinput! { enum Foo<'a> {} };
	assert_snapshot!(syntax, data, "enum_with_variable_count_type");
}

#[test]
fn bounded_range_multiplier_is_optimized_to_options() {
	let syntax = to_valuedef!(" <animateable-feature>{1,3} ");
	let data = to_deriveinput! { struct Foo; };
	assert_snapshot!(syntax, data, "bounded_range_multiplier_is_optimized_to_options");
}

#[test]
fn bounded_range_multiplier_is_optimized_to_options_with_lifetimes_when_necessary() {
	let syntax = to_valuedef!(" <'border-top-color'>{1,2} ");
	let data = to_deriveinput! { struct Foo<'a> {} }; // Foo specifies lifetime
	assert_snapshot!(syntax, data, "bounded_range_multiplier_is_optimized_to_options_with_lifetimes_when_necessary");
}

#[test]
fn value_fixed_range_color2_optimized() {
	let syntax = to_valuedef! { <color>{2} };
	let data = to_deriveinput! { struct Foo {} };
	assert_snapshot!(syntax, data, "value_fixed_range_color2_optimized");
}

#[test]
fn value_fixed_range_auto_color2_optimized() {
	let syntax = to_valuedef! { auto | <color>{2} };
	let data = to_deriveinput! { enum Foo {} };
	assert_snapshot!(syntax, data, "value_fixed_range_auto_color2_optimized");
}
