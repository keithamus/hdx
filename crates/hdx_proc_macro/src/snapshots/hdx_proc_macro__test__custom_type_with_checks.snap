---
source: crates/hdx_proc_macro/src/test.rs
expression: pretty
---
#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
#[cfg_attr(feature = "serde", derive(serde::Serialize), serde())]
enum Foo {
    None(::hdx_parser::T![Ident]),
    LengthPercentage(types::LengthPercentage),
}
#[automatically_derived]
impl<'a> ::hdx_parser::Peek<'a> for Foo {
    fn peek(p: &::hdx_parser::Parser<'a>) -> bool {
        use ::hdx_parser::Peek;
        p.peek::<::hdx_parser::T![Ident]>() || p.peek::<types::LengthPercentage>()
    }
}
#[automatically_derived]
impl<'a> ::hdx_parser::Parse<'a> for Foo {
    fn parse(p: &mut ::hdx_parser::Parser<'a>) -> ::hdx_parser::Result<Self> {
        use ::hdx_parser::Parse;
        if p.peek::<::hdx_parser::T![Ident]>() {
            let c = p.peek_n(1);
            match p.parse_atom_lower(c) {
                ::hdx_atom::atom!("none") => {
                    return Ok(Self::None(p.parse::<::hdx_parser::T![Ident]>()?));
                }
                atom => {}
            }
        }
        let start = p.offset();
        let val = p.parse::<types::LengthPercentage>()?;
        let valf32: f32 = val.into();
        if 0f32 > valf32 {
            return Err(
                ::hdx_parser::diagnostics::NumberTooSmall(
                    0f32,
                    ::hdx_lexer::Span::new(start, p.offset()),
                ),
            )?;
        }
        Ok(Self::LengthPercentage(val))
    }
}
#[automatically_derived]
impl ::hdx_parser::ToCursors for Foo {
    fn to_cursors(&self, s: &mut impl ::hdx_parser::CursorSink) {
        match self {
            Self::None(inner) => {
                s.append(inner.into());
            }
            Self::LengthPercentage(inner) => {
                ::hdx_parser::ToCursors::to_cursors(inner, s);
            }
        }
    }
}
